docs/dev/typesupport.md

```diff
@@ -183,3 +189,146 @@ union type {
 };
 ```
 
+#### Details
+
+Although the TypeTree follows the IDL syntax, there is one notable
+exceptions. Consider for example, the following fragment of IDL:
+
+```C
+struct s {
+   short a[3], b[5][6];
+};
+```
+
+The the syntax tree would look like:
+```
+   struct 's'
+   |
+   +-- member
+       |
+       +-- type 'short'
+       |
+       +-- declarators
+           |
+           +-- declarator 'a'
+           |   |
+           |   +-- array size '3'
+           |
+           +-- declarator 'b'
+               |
+               +-- array size '5'
+               |
+               +-- array size '6'
+```
+
+But in the TypeTree it is better to use the following tree representation
+that follows the 'semantic' representation of a type better:
+
+```
+   struct 's'
+   |
+   +-- declaration 'a'
+   |   |
+   |   +-- array size '3'
+   |       |
+   |       +-- short   <-------+
+   |                           |
+   +-- declaration 'b'         |
+       |                       |
+       +-- array size '5'      |
+           |                   |
+           +-- array size '6'  |
+               |               |
+               +---------------+
+```
+
+Note that the type 'short' is shared between the two declarations and
+that a far more complex type could have been used instead.
+During the freeing of the TypeTree, some mechanism is needed to determine
+if a tree is shared or not. Although, reference counting is commonly used
+for this, we decide to use a special flag for this. In the above example
```

Comment:
@FransFaase There is a reason reference counts are commonly used — could you please explain a little bit more about the reason why a single bit suffices?

E.g., is it that there is a clear notion of ownership, that all other references have a bit set to indicate that they are not, and that there is a mechanism that guarantees the owning references always outlasts all other references?

Reply:
The current implementation assumes that the TypeTree are constructed once, during parsing the IDL-grammar or, in the future, one of the representation described in DDS-XTYPES, and never modified with respect to their structure.
The functions defined in typetree.h for constructing the TypeTree still lack a description of the ownership rules. The implementation of the function (in typetree.c) currently lack any checks on incorrect usage. If you think it is worth the effort to add these, I am happy to add them.

Reply:
@FransFaase: I'd really appreciate it if you'd add a brief description of the ownership rules. Checking for usage ... I'd say that's worth it if it is a small amount of work.

Reply:
I have added some documentation to the functions in typetree.h
Every ddsts_type_t struct can be at most owned by one other ddsts_type_t. And if and only if, the struct is owned than the 'parent' member of ddsts_typespec_t points to its owner. The ownership relation between ddsts_type_t is a tree, where the root is the only struct with 'parent' equal NULL. When deleting the tree (from the root), it is sufficient to have only one flag, to mark all references that are not ownership relations, to take care that all structs are freed exactly once.

Reply:
Thanks @FransFaase. So that leaves only being unable to build on macOS and Windows. Would you be able to make it build on those platforms? It looks like it's simply a dependency on the "ddsc" library when generating the "ddsts" library.

Reply:
@FransFaase, I looked at the build problem a little bit better, and everything indeed points to the problem being that it doesn't link with "ddsc" when generating the "ddsts" shared library. While on Linux undefined symbols in a shared library is acceptable under the default settings, on macOS and Windows it isn't.

This in turn matches src/ddsts/CMakeLists.txt: there is no dependency on ddsc. Adding
```CMake
target_link_libraries(ddsts PRIVATE ddsc)
```
solves the problem on macOS but my CMake knowledge is entirely insufficient to know whether this introduces all kinds of other hidden issues. @k0ekk0ek, perhaps you can shed some light on this?

Reply:
@eboasson, yes, that seems to be the solution, because ddsrt is made available in the ddsc (dynamic) library.